File: day1/javascript/solution2.js
//advent of code day 1
import fs from 'node:fs'
import readLine from 'node:readline'

//get params from command line
const args = process.argv.slice(2)
const filename = args[0]

const fileStream = fs.createReadStream(filename)
const rl = readLine.createInterface({
  input: fileStream,
  crLfDelay: Infinity
})

let totalSum = 0;
const numbers = new Map(Object.entries({
  "one": 1,
  "two": 2,
  "three": 3,
  "four": 4,
  "five": 5,
  "six": 6,
  "seven": 7,
  "eight": 8,
  "nine": 9
}))

rl.on('line', (line) => {
  processLine(line)
})

rl.on('close', () => {
  console.log(`totalSum: ${totalSum}`)
})

function reverseString(str) {
  return str.split("").reverse().join("");
}

function processLine(line) {
  let firstNumber;
  let lastNumber;
  let posibleNumber = ""
  let sum = 0;
  let found = false

  for (let n of line) {
    if (found) break

    if (Number.isInteger(parseInt(n))) {
      firstNumber = n
      found = true
    }

    posibleNumber += n
    if (posibleNumber.length >= 3) {
      for (let key of numbers.keys()) {
        if (posibleNumber.indexOf(key) !== -1) {
          firstNumber = numbers.get(key)
          found = true
          break
        }
      }
    }
  }

  posibleNumber = ""
  found = false

  for (let i = line.length; i >= 0; i--) {
    if (found) break

    if (Number.isInteger(parseInt(line[i]))) {
      lastNumber = line[i]
      found = true
    }

    if (posibleNumber.length >= 3) {

      for (let key of numbers.keys()) {

        if (posibleNumber.indexOf(reverseString(key)) !== -1) {
          lastNumber = numbers.get(key)
          found = true
          break
        }
      }
    }
    posibleNumber += line[i]
  }

  sum = String(firstNumber).concat(lastNumber)
  totalSum += parseInt(sum)
  console.log(`first: ${firstNumber}, last: ${lastNumber}, sum: ${sum}, totalSum: ${totalSum} line: ${line}`)
}

File: day1/c/solution2.c
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, char *argv[]) {
  long int totalSum = 0;
  long int sum = 0;
  char lineStr[100];
  char reversedLineStr[100];
  char concatNumber[20];
  char firstNumber[2];
  char lastNumber[2];
  char *fileName = argv[1];
  char posibleNumber[90];
  FILE *fptr;

  // Char type is not a null terminated string, so we need to add the null
  firstNumber[1] = '\0';
  lastNumber[1] = '\0';

  // use appropriate location if you are using MacOS or Linux
  if ((fptr = fopen(fileName, "r")) == NULL) {
    // Program exits if the file pointer returns NULL.
    exit(1);
  }

  while (fgets(lineStr, 100, fptr) != NULL) {
    for (int i = 0; i < strlen(lineStr); i++) {

      if (isdigit(lineStr[i])) {
        firstNumber[0] = lineStr[i];
        break;
      }

      strncat(posibleNumber, &lineStr[i], 1);
      if (strlen(posibleNumber) >= 3) {

        if (strstr(posibleNumber, "one") != NULL) {
          firstNumber[0] = '1';
          break;
        }
        if (strstr(posibleNumber, "two") != NULL) {
          firstNumber[0] = '2';
          break;
        }
        if (strstr(posibleNumber, "three") != NULL) {
          firstNumber[0] = '3';
          break;
        }
        if (strstr(posibleNumber, "four") != NULL) {
          firstNumber[0] = '4';
          break;
        }
        if (strstr(posibleNumber, "five") != NULL) {
          firstNumber[0] = '5';
          break;
        }
        if (strstr(posibleNumber, "six") != NULL) {
          firstNumber[0] = '6';
          break;
        }
        if (strstr(posibleNumber, "seven") != NULL) {
          firstNumber[0] = '7';
          break;
        }
        if (strstr(posibleNumber, "eight") != NULL) {
          firstNumber[0] = '8';
          break;
        }
        if (strstr(posibleNumber, "nine") != NULL) {
          firstNumber[0] = '9';
          break;
        }
      }
    }
    // strcpy(reversedLineStr, lineStr);
    // strrev(reversedLineStr);
    strcpy(posibleNumber, "");
    for (int i = (strlen(lineStr) - 1); i >= 0; i--) {
      if (isdigit(lineStr[i])) {
        lastNumber[0] = lineStr[i];
        break;
      }

      strncat(posibleNumber, &lineStr[i], 1);
      if (strlen(posibleNumber) >= 3) {

        if (strstr(posibleNumber, "eno") != NULL) {
          lastNumber[0] = '1';
          break;
        }
        if (strstr(posibleNumber, "owt") != NULL) {
          lastNumber[0] = '2';
          break;
        }
        if (strstr(posibleNumber, "eerht") != NULL) {
          lastNumber[0] = '3';
          break;
        }
        if (strstr(posibleNumber, "ruof") != NULL) {
          lastNumber[0] = '4';
          break;
        }
        if (strstr(posibleNumber, "evif") != NULL) {
          lastNumber[0] = '5';
          break;
        }
        if (strstr(posibleNumber, "xis") != NULL) {
          lastNumber[0] = '6';
          break;
        }
        if (strstr(posibleNumber, "neves") != NULL) {
          lastNumber[0] = '7';
          break;
        }
        if (strstr(posibleNumber, "thgie") != NULL) {
          lastNumber[0] = '8';
          break;
        }
        if (strstr(posibleNumber, "enin") != NULL) {
          lastNumber[0] = '9';
          break;
        }
      }
    }
    // printf("Concatenated number before Concatenation: %s\n", concatNumber);
    strcat(concatNumber, firstNumber);
    // printf("Concatenated number after first Concatenation: %s\n",
    // concatNumber); printf("Lastnumber value: %s\n", lastNumber);
    strcat(concatNumber, lastNumber);
    // printf("Concatenated number: %s %p\n", concatNumber, &concatNumber);

    sum = strtol(concatNumber, NULL, 10);
    totalSum = totalSum + sum;
    printf("first: %c last: %c suma: %ld totalAcc: %ld cadena: %s",
           firstNumber[0], lastNumber[0], sum, totalSum, lineStr);
    strcpy(concatNumber, "");
    strcpy(posibleNumber, "");
  }

  fclose(fptr);
  printf("La Suma total es %ld\n", totalSum);

  return 0;
}
